VC-CodeLabs/russian_nesting
# JeffR: Russian Nesting Solution

I wanted to use modules, so needed a go.mod, so isolated everything to a JeffR/ folder

the libsln/ folder contains the solution implementation; the code that finds the max nested envelopes is in solution.go- once we have the envelope collection loaded, we find the max nested envelopes by calling `FindMaxNestedEnvelopes(CompactEnvelopes(SortEnvelopes(envelopes)))`; interfaces were used to support swapping in different solution impls during development.

to test input from a file:
e.g.
`go run JeffR_RussianNesting_Solution.go < samples/JeffR_Alek_Example1.txt`
(flip the slash for windows cmd)
additional notes on piping file content or manual input for testing are provided in my root readme.

samples/testSamples.go provides a basic test harness for validating all the samples driven by testCases*.txt- see the readme for add'l details

the test folder should be ignored, it was really more R&D

*****

Update 4/22 a.m.:
I took advantage of the more liberal deadline (_"if you submit early Monday, that's okay" --Alek in some fairly recent CodeLabs review meeting_)

Added a max "chunked" sample which revealed how slow it can be vs the max "all unique" and max "all the same" samples, where the former requires more labor vs the latter two take advantage of optimizations to short-circuit the process; so, I added threading support, which as usual is a bit of a mixed bag- max "unique" got a little worse, max "all match" got slightly worse while max "chunked" got _much_ better- see the readme in test/ for additional details.

Whether or not threading was the right call will depend mostly on the actual test cases.

****

Update 4/23 a.m.

General cleanup, refined some names and added/updated code comments
